third_party/qemu/README.md

```diff
@@ -1,48 +1,112 @@
+
+Currently, only the Earlgrey machine (`ot-earlgrey`) has any support within OpenTitanLib for test integration, primarily on the `earlgrey_1.0.0` branch. As of the time of writing, a **non-exhaustive** list of some of the major features not emulated by QEMU include:
+* Mostly or entirely unimplemented devices: AST, pattgen, PWM, sensor_ctrl, sysrst_ctrl, usbdev.
+* Power manager slow FSM is bypassed, with no support for low power entry / deep sleeps.
+* A pinmux block is implemented, but all connections currently go via a CharDev and do not interact with pinmux.
+* Non-Maskable Interrupts (NMIs) are not implemented due to Ibex's use of non-standard NMIs.
+* Key masking (AES, KMAC, etc.) is not implemented.
+* Entropy FSMs & scheduling for entropy consumers (e.g. KMAC, keymgr) are broadly unimplemented.
+* Synchronous (e.g. I2C) or timing-based (UART) communication peripherals are often limited in support due to the inherent inaccuracy of QEMU's emulation.
+* Many other implemented features across all IPs, such as flash ECCs & scrambling, clkmgr gating and jitter, aon_timer pauses, alert emulation for several devices, etc.
+
+## Considerations & Constraints
+
+### Timing and Cycle Accuracy
+
+QEMU is _not_ a cycle-accurate emulator, as its main goal is emulation speed.
+The instruction count reported in `mcycle` will not accurately reflect the number of instructions that have been executed.
+When using a hardware timer (e.g. the `aon_timer` or `rv_timer`), the timer will expire after _approximately_ the correct amount of time, but will not be exactly accurate.
+
+QEMU has mechanisms for pacing the virtual CPU to match hardware timers, but in general it should be expected that both hardware timers and instruction counts are not correct in QEMU.
+This can be a particular challenge for security countermeasures and software mechanisms that opt to set tight timing thresholds.
+One potential solution is modifying the `icount` shift parameter, which informs the TCG's instruction counting feature for system emulation.
+When providing an icount shift of `N`, the virtual CPU will execute one instruction every `2^N` nanoseconds of virtual time, meaning that changing this parameter may allow emulation to "pass" tighter timing checks at the cost of less accurate timing overall.
+By default, an `icount` of 6 is used to align execution time with wall-clock time.
+
+### The TLB and ePMP
+
+Hardware will often use a Translation Lookaside Buffer (TLB) as a high-speed cache to accelerate translation from virtual addresses to physical addresses.
+QEMU mimics this concept with its own address translation cache that it calls the [TLB](https://www.qemu.org/docs/master/devel/tcg.html#mmu-emulation), which speeds up translation from guest virtual addresses to either virtual addresses of host RAM or MMIO operations for emulated devices.
```

Comment:
Guest Virtual to Host Virtual is true, but the equivalent which I think makes more sense here is Guest Physical.
```suggestion
QEMU mimics this concept with its own address translation cache that it calls the [TLB](https://www.qemu.org/docs/master/devel/tcg.html#mmu-emulation), which it uses to speed up translation of guest virtual addresses to guest RAM or to MMIO regions of emulated devices.
```

Reply:
AFAIK Your understanding is correct, but I think the data and code structure of the QEMU TLB is an implementation detail that I do not think needs much elaborating, just that it performs the functions of an MMU in translating Guest Virtual to Guest Phyiscal addresses, which could be in plain memory or memory-mapped regions that dispatch to I/O.